lib/valkyrie/shrine/storage/s3.rb

```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+require "shrine/storage/s3"
+
+module Valkyrie
+  module Shrine
+    module Storage
+      class S3 < ::Shrine::Storage::S3
+        # List objects id's that are starting with a given prefix.
+        # This is helpful for versioned files that have a common file identifier.
+        # Need to make sure to combine with storage prefix.
+        #    list_object_ids(id_prefix: "some/object/id")
+        # @param id_prefix [String] - object's id that starts with
+        # @return [Array(String)]
+        def list_object_ids(id_prefix:)
+          bucket.objects(prefix: [*prefix, id_prefix].join("/"))
+                .map { |obj| (prefix.present? ? obj.key.sub(/^#{prefix}\//, "") : obj.key) }
```

Comment:
I would keep this on `Valkyrie::Storage::VersionedShrine` rather than subclass `S3`.

Also, I think this can probably be implemented more cleanly with [`AWS::S3::Client#list_objects_v2`](https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/S3/Client.html#list_objects_v2-instance_method), which you can get from `Shrine::Storage::S3#client`.

Reply:
I don't think AWS::S3::Client#list_objects_v2 can make it simpler since we need to handle the raw AWS response with pagination, and we might still need to handle the shrine specific issues like prefix.

It looks more like shrine should encapsulate things like bucket, prefix etc. that ShrineVersion doesn't need to know about. But we can move it to a private method in Valkyrie::Storage::VersionedShrine if you like, or implemented inside #version_files, which might not be a big deal though.

Reply:
Sure. Let's make it.